### PR TITLE
custom embeddings

### PR DESCRIPTION
**Pull request by tsinggggg**
1. extended the WordEmbedding class to implement 2 ways to initialize a word embedding: one is as before downloading from s3 all the files needed(embeddingmat, cos_sim_mat, etc..); the other being using gensim keyedvector to avoid re-implementing things(example in test case) 
2. incorporated WordEmbeddings with WordSwapEmbedding and WordEmbeddingDistance and ThoughtVector as per https://github.com/QData/TextAttack/issues/325#issuecomment-724888184. APIs of these 3 class are unchanged except an additional argument in __init__ to indicate whether to use embeddings from textattack s3 or in gensim keyedvectors format.

### CODE REVIEW

**Comment by jinyongyoo** [comment]
@tsinggggg Thanks for the PR! Do you mind if I make some changes too? I like how you integrated gensim embeddings, but I also think there's a more general way we can set up the `WordEmbedding` class so it can work with other custom embeddings.

**Comment by tsinggggg** [comment]
> @tsinggggg Thanks for the PR! Do you mind if I make some changes too? I like how you integrated gensim embeddings, but I also think there's a more general way we can set up the `WordEmbedding` class so it can work with other custom embeddings.

Not at all. I'd love to see a more general solution as well! @jinyongyoo

**Comment by jinyongyoo** [comment]
@tsinggggg 

I felt that the API for creating `WordEmbedding` object was unclear, and it was even less clear if the user wanted to use their own word embedding that might be a different format. So I made `WordEmbedding` a base class so users know exactly what methods they to implement to use their own custom word embeddings. Also `TextAttackWordEmbedding` implements most of the necessary functions, so users can plug in appropriate components and have a word embedding class that works without implementing their own. Lastly `GensimWordEmbedding` provides a clear way to use gensim's `KeyedVectors`. Please let me know if you have any questions.

@jxmorris12 
Can you review this? I was planning to review it, but I ended up writing new stuff myself.

**Comment by qiyanjun** [comment]
@jxmorris12 can you review?

File: textattack/shared/word_embedding.py [hunk 1]
```diff
@@ -3,103 +3,437 @@
...
+            neighbours (list[int]): List of indices of the nearest neighbours
+        """
+        raise NotImplementedError()
+
+    __repr__ = __str__ = utils.default_class_repr
+
+
+class TextAttackWordEmbedding(WordEmbedding):
+    """Object for loading word embeddings and related distances for TextAttack.
+    This class has a lot of internal components (e.g. get consine similarity)
```
Comment (jxmorris12) [inline textattack/shared/word_embedding.py:105]: typo: cosine

**Review by jxmorris12 — APPROVED** [review]
LGTM, but it's not immediately obvious what the difference between `WordEmbedding` and  `TextAttackWordEmbedding` is, or why they're separate classes in the first place.

**Comment by jinyongyoo** [comment]
> LGTM, but it's not immediately obvious what the difference between `WordEmbedding` and `TextAttackWordEmbedding` is, or why they're separate classes in the first place.

Well `WordEmbedding` defines what methods need to implemented if a user wants to come up with a custom word embedding class. `TextAttackWordEmbedding` is the class that actually has a lot of the internal stuff implemented and just requires the user to pass in the appropriate matrices and dictionary. Most users should just use `TextAttackWordEmbedding`, but `WordEmbedding` is there just in case some really wants to implement their own methods.

**Comment by jxmorris12** [comment]
> > LGTM, but it's not immediately obvious what the difference between `WordEmbedding` and `TextAttackWordEmbedding` is, or why they're separate classes in the first place.
> 
> Well `WordEmbedding` defines what methods need to implemented if a user wants to come up with a custom word embedding class. `TextAttackWordEmbedding` is the class that actually has a lot of the internal stuff implemented and just requires the user to pass in the appropriate matrices and dictionary. Most users should just use `TextAttackWordEmbedding`, but `WordEmbedding` is there just in case some really wants to implement their own methods.

I see. That makes sense. It seems a little arbitrary which one is which name-wise, is all.

**Pull request merged**